File: Programmers/greedy1.py
def solution(n, lost, reserve):
    answer = n

    _lost = [l for l in lost if l not in reserve]
    _reserve = [r for r in reserve if r not in lost]

    for l in _lost:
        if l - 1 in _reserve:
            _reserve.remove(l - 1)
        elif l + 1 in _reserve:
            _reserve.remove(l + 1)
        else:
            answer -= 1

    return answer

n = 5
lost = [2, 4]
reserve = [1, 3, 5]
# answer = 5
print(solution(n, lost, reserve))

n = 5
lost = [2, 4]
reserve = [3]
# answer = 4
print(solution(n, lost, reserve))

n = 3
lost = [3]
reserve = [1]
# answer = 2
print(solution(n, lost, reserve))

File: Programmers/greedy1.js
const mergeSort = (array) => {
  if (array.length < 2) return array;
  const pivot = Math.floor(array.length / 2);
  const left = array.slice(0, pivot);
  const right = array.slice(pivot, array.length);
  return merge(mergeSort(left), mergeSort(right));
};

const merge = (left, right) => {
  const result = [];
  while (left.length && right.length) {
    if (left[0] <= right[0]) {
      result.push(left.shift());
    } else {
      result.push(right.shift());
    }
  }

  while (left.length) result.push(left.shift());
  while (right.length) result.push(right.shift());
  return result;
};

function solution(n, lost, reserve) {
  var answer = n;

  const _lost = lost.filter((l) => !reserve.includes(l));
  const _reserve = reserve.filter((r) => !lost.includes(r));

  mergeSort(_lost);
  const nonattendance = _lost.filter((l) => {
    const index = _reserve.findIndex((r) => r == l - 1 || r == l + 1);
    if (index >= 0) {
      _reserve.splice(index, 1);
      return false;
    } else {
      return true;
    }
  }).length;
  return answer - nonattendance;
}